src/Lean/Server/FileWorker/RequestHandling.lean

```diff
@@ -133,54 +135,66 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
               return ← ci.runMetaM i.lctx <| locationLinksFromDecl i ei.elaborator
       return #[]
 
-open Elab in
-partial def handlePlainGoal (p : PlainGoalParams)
-    : RequestM (RequestTask (Option PlainGoal)) := do
+open RequestM in
+def getInteractiveGoals (p : Lsp.PlainGoalParams) : RequestM (RequestTask Widget.InteractiveGoals) := do
   let doc ← readDoc
   let text := doc.meta.text
   let hoverPos := text.lspPosToUtf8Pos p.position
   -- NOTE: use `>=` since the cursor can be *after* the input
   withWaitFindSnap doc (fun s => s.endPos >= hoverPos)
-    (notFoundX := return none) fun snap => do
+    (notFoundX := return { goals := #[] }) fun snap => do
       for t in snap.cmdState.infoState.trees do
         if let rs@(_ :: _) := t.goalsAt? doc.meta.text hoverPos then
           let goals ← List.join <$> rs.mapM fun { ctxInfo := ci, tacticInfo := ti, useAfter := useAfter } =>
             let ci := if useAfter then { ci with mctx := ti.mctxAfter } else { ci with mctx := ti.mctxBefore }
             let goals := if useAfter then ti.goalsAfter else ti.goalsBefore
-            ci.runMetaM {} <| goals.mapM (fun g => Meta.withPPInaccessibleNames (Meta.ppGoal g))
-          let md :=
-            if goals.isEmpty then
-              "no goals"
-            else
-              let goals := goals.map fun goal => s!"```lean
-{goal}
-```"
-              String.intercalate "\n---\n" goals
-          return some { goals := goals.map toString |>.toArray, rendered := md }
+            ci.runMetaM {} <| goals.mapM (fun g => Meta.withPPInaccessibleNames (Widget.goalToInteractive g))
+          return { goals := goals.toArray }
 
-      return none
+      return { goals := #[] }
 
 open Elab in
-open Meta in
-partial def handlePlainTermGoal (p : PlainTermGoalParams)
-    : RequestM (RequestTask (Option PlainTermGoal)) := do
+partial def handlePlainGoal (p : PlainGoalParams)
+    : RequestM (RequestTask (Option PlainGoal)) := do
+  let t ← getInteractiveGoals p
+  t.map <| Except.map fun ⟨goals⟩ =>
+    if goals.isEmpty then
+      some { goals := #[], rendered := "no goals" }
```

Comment:
I'll look at this branch more closely later.  However this is a breaking change: it now shows "goals accomplished" everywhere (even when there's no tactic proof).

Reply:
Ah thanks, that's a bug. I forgot about the `none` meaning not a proof and `#[]` meaning goals accomplished semantics.